### ISSUE TITLE
Avoid magic strings in IModule

### ISSUE THREAD

**Issue by hishamco** [oqtane/oqtane.framework#322]
As I have seen that `IModule` contains a `Properties` which is a set of module properties that the Oqtane will use, but the problem is they are magic string, so:

- User may miss some of the properties

- Casing may be another issue

- Spelling is another issue

to avoid such things I suggest to introduce a `ModuleAttribute` or `ModuleMaifestAttribute` that contains a set of all the required attributes that we have, so:

- It's strongly typed

- Will fix all of the above issues

- It will be useful in modules discovery if the modules are come from other assemblies

I can work for a prototype for this is sound looks good to you

/cc @sbwalker

**Comment by sbwalker** [comment]
Please do not make this change. The original prototype for IModule was a strongly typed object. However as soon as you add an addition property to a strongly typed object it breaks backward compatibility with any module which was built against the prior version of the interface. This is not acceptable as it breaks then entire extensibility. Using a generic Dictionary allows new properties to be added while still maintaining backward compatibility.

**Comment by sbwalker** [comment]
If you would like to introduce a class which defines constants for the list of supported dictionary keys it would solve the 3 problems you identified, and still ensure the interface maintains compatibility.

**Comment by hishamco** [comment]
> The original prototype for IModule was a strongly typed object

The object is strongly typed while the properties is not

While we don't ship `1.0.0` breaking changes may still valid if I'm not wrong, specially if it's crucial, to maintain the compatibility we can introduce the attribute without remove `Properties` property, and we can mark the properties as `Obsolete`, so both should run fine

> If you would like to introduce a class which defines constants for the list of supported dictionary keys it would solve the 3 problems you identified, and still ensure the interface maintains compatibility.

Even though this will not solve the problem unless the module developer knows about the class that contains the keys

**Comment by hishamco** [comment]
I can make a prototype for this and still maintain the backward compatibility

**Comment by sbwalker** [comment]
Originally IModule was a strongly typed object which a property for each setting rather than a Dictionary. I then created the Blog module project and I implemented the IModule interface. This worked fine. However then I added an additional property to the IModule interface. And when I tried to run the Blog module it threw an error because it did not implement the new version of the IModule interface.

New requirements are going to materialize so you can't assume that an interface like this is going to remain static. In DNN we used an external XML to define the module manifest - which obviously worked fine as requirements changed, as there was not a code level coupling. However I want to avoid external files in Oqtane as there are a variety of other problems with this approach.

**Comment by sbwalker** [comment]
Maybe if you provided an example of the interface definition it would help me understand what you are saying

**Comment by hishamco** [comment]
I will not introduce an interface, probably something like `ModuleManifestAttribute`

```csharp
public class ModuleManifestAttribute : Attribute
{
    public ModuleManifestAttribute(string name, string author, string description, string version)
    {
        ...
    }

    public string Name { get; }
   
    public string Author { get; }

    public string Description { get; }

    public string Version { get; }
}
```

**Comment by sbwalker** [comment]
This was the original IModule implementation:

namespace Oqtane.Modules
{
    public interface IModule
    {
        string Name { get; }
        string Description { get; }
        string Version { get; }
        string Owner { get; }
        string Url { get; }
        string Contact { get; }
        string License { get; }
        string Dependencies { get; }
    }
}

and an example implementation:

namespace Oqtane.Client.Modules.Counter
{
    public class Module : IModule
    {
        public string Name { get { return "Counter"; } }
        public string Description { get { return "Increments a counter"; } }
        public string Version { get { return "1.0.0"; } }
        public string Owner { get { return ""; } }
        public string Url { get { return ""; } }
        public string Contact { get { return ""; } }
        public string License { get { return ""; } }
        public string Dependencies { get { return ""; } }
    }
}

I think what you are saying is use an interface with an object rather than properties: 

ie.

namespace Oqtane.Modules
{
    public interface IModule
    {
        ModuleDefinition ModuleDefinition { get; }
    }
}

**Comment by sbwalker** [comment]
I dont understand how an attribute would be applicable in this case as the IModule properties describe the entire module - not just a single razor class

**Comment by hishamco** [comment]
I know, let me give you a proper example:

Counter module will have

```csharp
[assembly: Module(
    Name = "Counter",
    Author = "The Oqtane Team",
    Version = "1.0.0",
    Description = "Increments a counter"
)]
```

so every module in the system will have it's own manifest that describe all sets of module properties, then in module discovery we will look for that attribute to get all the modules definitions

**Comment by sbwalker** [comment]
In the current system, a module does not need to implement IModule - this was on purpose to make it simpler for developers

**Comment by hishamco** [comment]
In the source I just find something like this

```csharp
    public class Module : IModule
    {
        public Dictionary<string, string> Properties
        {
            get
            {
                Dictionary<string, string> properties = new Dictionary<string, string>
                {
                    { "Name", "Counter" },
                    { "Description", "Increments a counter" },
                    { "Version", "1.0.0" }
                };
                return properties;
            }
        }
    }
```

**Comment by hishamco** [comment]
> For example, all of the Admin modules do not have an IModule interface. This was the philosophy of "Less Ceremony"

How the system will know about the module metadata?

**Comment by sbwalker** [comment]
Some modules do not need to provide metadata - the metadata is inferred based on the class name

**Comment by hishamco** [comment]
AFAIK each module should have sort of metadata so other module can know about them, and more important the system show know, in case to enable / disable them where it's needed

**Comment by sbwalker** [comment]
Its not required. The IModule interface is currently optional. If the IModule is not provided for a module it uses the class name to set some default metadata for the module. The only critical metadata for a module is the typename as this is how the module is dynamically instantiated. I do not see any valid reason to change the existing functionality.

**Comment by sbwalker** [comment]
The focus of this particular issue is to consider changing the IModule interface definition. If backward compatibility can be preserved then I am open to the change. But I do not want to change the current behavior as it relates to module loading and the IModule interface.

**Comment by hishamco** [comment]
Let me create a prototype and share it with you ...

**Comment by sbwalker** [comment]
Yes, please create a prototype. I was just trying to provide some guidance so that you do not spend a lot of time and effort on something that I will not be able to accept as a contribution. I do think that there may be a way to modify the IModule interface so that it maintains backward compatibility and achieves the goals you outlined above. I do not want to refactor the module loading logic for the reasons I provided above as well as the fact as this logic is working and is also consistent with the way that themes are currently managed as well.

**Comment by hishamco** [comment]
Altering the `IModule` may introduce a breaking change, so that's why I suggest to create a new attribute

BTW why I don't see any module using `ModuleDefinition` which contains most of the properties I need to introduce in the proposed attribute?!!

**Comment by sbwalker** [comment]
The IModule interface is required for current module loading process, so please do not use an Attribute.

**Comment by hishamco** [comment]
My aim to make both work, otherwise changing the `IModule` may introduce breaking changes

**Comment by sbwalker** [comment]
The Module entity has a ModuleDefinition property which is populated in the ModuleController. The properties in ModuleDefinition are referenced in many places in modules.

**Comment by sbwalker** [comment]
If you are saying that you plan to make both the attribute and interface work, I would recommend not going down that path as we do not want to have 2 methods for developers to provide metadata for modules - this will result in more framework code to test/maintain in the long term. Plus it breaks one of the fundamental principles of Oqtane which is to keep things very simple for developers.

**Comment by hishamco** [comment]
I saw the references in controllers and repositories, anyhow while not all the modules have a definitions in the code as I expected no need to doing this change, if the attribute not make sense to you

Thanks

**Comment by hishamco** [comment]
> If you are saying that you plan to make both the attribute and interface work,

That will happen to make the `IModule` backward compatible, which is I think it's fine to break the APIs before `1.0.0` release

**Comment by sbwalker** [comment]
Yes, if we want to make an interface change it is better to do it < 1.0.0. I do believe that your earlier comment related to interfaces "The object is strongly typed while the properties is not" is accurate. So changing the IModule interface to something like the following may actually work ( and would achieve the 3 goals you outlined )

namespace Oqtane.Modules
{
public interface IModule
{
ModuleDefinition ModuleDefinition { get; }
}
}

with an implementation such as:

   public class Module : IModule
    {
        public ModuleDefinition ModuleDefinition
        {
            get
            {
                return new ModuleDefinition
                {
                    Name = Counter",
                    Description = "Increments a counter",
                    Version = "1.0.0"
                };
            }
        }
    }

**Comment by hishamco** [comment]
This is another solution which fixes the above issues, if it's fine I can send a PR

**Comment by sbwalker** [comment]
Yes please. And if we can prove this approach preserves backward compatibility then perhaps you can also refactor ITheme to use the same approach. I do prefer strongly typed properties to a generic dictionary.

**Comment by hishamco** [comment]
Sure ..

**Comment by sbwalker** [comment]
Resolved in #343, #344
